### DISCUSSION TITLE
MariaDB permissions issues when doing fresh install

### DISCUSSION THREAD

**Discussion by captainkev76** [photoprism/photoprism#1652] [Q&A]
Hi All,

I have VirtualBox on Windows 10, and I have created an Ubuntu 20.04 server VM, and installed Docker on it.

I'm trying to get the Photoprism docker container working, but keep getting problems when it tries to start up MariaDB:

`mariadb_1     | 2021-10-22 18:48:38+00:00 [Note] [Entrypoint]: Entrypoint script for MariaDB Server 1:10.5.12+maria~focal started.
mariadb_1     | 2021-10-22 18:48:38+00:00 [Note] [Entrypoint]: Switching to dedicated user 'mysql'
mariadb_1     | 2021-10-22 18:48:38+00:00 [Note] [Entrypoint]: Entrypoint script for MariaDB Server 1:10.5.12+maria~focal started.
mariadb_1     | 2021-10-22 18:48:38+00:00 [Note] [Entrypoint]: Initializing database files
mariadb_1     | 2021-10-22 18:48:38 0 [Warning] Can't create test file /var/lib/mysql/7dc273281465.lower-test
mariadb_1     | /usr/sbin/mysqld: Can't change dir to '/var/lib/mysql/' (Errcode: 13 "Permission denied")
mariadb_1     | 2021-10-22 18:48:38 0 [ERROR] Aborting
`

I'm struggling to understand why it's having this problem. Has anyone else come across this?

Kevin

**Comment by lastzero** [comment]
Did you use our config example for Windows?

https://dl.photoprism.org/docker/windows/docker-compose.yml

It uses a named volume, which Docker manages. Shouldn't cause any issues. Since you're using a VM, you may have used our regular config example for Linux: https://dl.photoprism.org/docker/docker-compose.yml

If so, the database stores its files in the `database` folder:

```
volumes:
      - "./database:/var/lib/mysql" # Never remove
```

It's in a sub folder of where the docker-compose.yml file is. Can you check if the directory exists and what permissions it has? Run `chmod -R 777 database` to give all users write permissions. Since it's all running inside a VM, this shouldn't be a security issue.

**Reply by captainkev76** [reply]
Thanks for the speedy reply! I did re-use the original docker-compose file from when it was installed in Docker for Desktop in Windows - which was indeed based on the config example for Windows.

I've checked and the database folder does exist. I ran the 'chmod' command as suggested, but the outcome is exactly the same as before.

I've also tried to create a completely fresh instance of Photoprism by bringing down the previous containers, doing a 'docker volume prune', and then creating a fresh folder with a fresh docker-compose.yml file. But this also has the same problem as before...

```
mariadb      | 2021-10-23 08:34:32+00:00 [Note] [Entrypoint]: Entrypoint script for MariaDB Server 1:10.5.12+maria~focal started.
mariadb       | 2021-10-23 08:34:32+00:00 [Note] [Entrypoint]: Switching to dedicated user 'mysql'
mariadb       | 2021-10-23 08:34:32+00:00 [Note] [Entrypoint]: Entrypoint script for MariaDB Server 1:10.5.12+maria~focal started.
mariadb       | 2021-10-23 08:34:33+00:00 [Note] [Entrypoint]: Initializing database files
mariadb       | 2021-10-23  8:34:33 0 [Warning] Can't create test file /var/lib/mysql/9819a8ccf918.lower-test
mariadb       | /usr/sbin/mysqld: Can't change dir to '/var/lib/mysql/' (Errcode: 13 "Permission denied")
mariadb       | 2021-10-23  8:34:33 0 [ERROR] Aborting
mariadb       |
mariadb       | Installation of system tables failed! 

```

[docker-compose-REDACTED.yml.TXT](https://github.com/photoprism/photoprism/files/7402879/docker-compose-REDACTED.yml.TXT)

**Reply by lastzero** [reply]
Does it help to change the folder name e.g. `"./mariadb:/var/lib/mysql"`?

If you're in fact directly or indirectly mounting a folder on your Windows host, you may have to change the permissions directly in Windows.

If nothing helps, the problem could also be a Linux kernel security module like SELinux (although I never had issues with that on Ubuntu 20.04):
- https://mariadb.com/kb/en/selinux/
- https://docs.photoprism.org/getting-started/troubleshooting/#common-issues

**Reply by captainkev76** [reply]
It doesn't seem to make any difference if I change the folder name to./mariadb. I've changed the permissions in Windows to give everyone access. I think it definitely does have access, as it is able to create the data and database/mariadb folders, although it seems it can't actually create files in there. 

I'm assuming MariaDB launches with UID:GID 1000:1000? This should have the necessary permissions but one of my other containers was also having problems with permissions, but when I ran the container as root (0:0) it worked fine, so I think something similar might be going on here. Is there a way to modify the docker-compose.yml file so that MariaDB launches with UID:GID 0:0? Not as a permanent solution, just to see if this does fix it or not. Thanks again for your help!

**Reply by lastzero** [reply]
Should start as root by default, but you can explicitly set `user: 0:0` for the mariadb service.

**Reply by lastzero** [reply]
You can also try the `mysql:8` image to see if it works better in your VM.

**Reply by captainkev76** [reply]
Still no luck with finding out what's going on here. I tried modifying the docker-compose file so that the database was created locally instead of on the network shared drive. This worked perfectly, but I'd like the database to be on the shared drive so that it's backed up by default. So there's obviously a permissions problem on the shared drive

I managed to find out that MariaDB creates user 'mysql', with UID:GID 999:999 to do the things it needs to do when setting up the database and tables. However, in Ubuntu, UID:GID 999:999 is assigned to systemd-coredump. Is this to be expected? 

When I done a chown -R 999:999 ./ in the shared drive folder where the database folder exists, still the same MariaDB error appeared.

Feel like I'm very close to cracking this conundrum, but it's still just beyond me...

Any ideas/thoughts would be most welcome :-)

**Reply by lastzero** [reply]
You didn't put the database folder inside the PhotoPrism storage folder, right? PhotoPrism might change permissions there if it runs not as root. You've also restarted properly using `docker-compose stop` and `docker-compose up` (not using `start`)? You said you tried this already, but in case not then try using a named volume for the database as shown in our [Windows example](https://dl.photoprism.org/docker/windows/docker-compose.yml):

```
  mariadb:
    restart: unless-stopped
    image: mariadb:10.6
[...]
    volumes:
      - "mariadb_data:/var/lib/mysql" # Never remove

[...]
     
volumes:
  mariadb_data:
    driver: local
```

**Reply by captainkev76** [reply]
The Photoprism meta data folder is on the U drive, do I've mounted it as /media/u/, and the storage folder (with the original photos) is on the M drive, so mounted as /media/m/. The volumes called in the docker-compose file are: /media/u/Docker/photoprism and /media/m/Photos/ respectively.

Using the named volume as suggested above works perfectly, but it means that the meta data folder is held locally in the docker container (I think?) and not on the shared drive where it would be auto backed up. I'm also now very curious as to why the shared folder is proving to be so problematic.

**Reply by lastzero** [reply]
Yes, it should be in the VM somewhere in `/var/...` when using named volumes (advanced users can change it in the Docker config). Can you manually create files in the folder where you expect the database to create them?

```
mkdir database
touch database/example.txt
```

Not sure what technology your VM is based on... When using Docker, it seems necessary to disable the WSL 2 based engine in Docker Settings > General to mount drives other than `C:` (at least that was the case when I tested it).  This will enable Hyper-V, which Microsoft doesn't offer to its Windows Home customers.

**Reply by captainkev76** [reply]
Just tried the mkdir and touch commands - both successful! I'm using Virtualbox on Windows, with an Ubuntu VM running in Virtualbox. and Docker running in Ubuntu. I believe Virtualbox uses VTx for virtualisation.

There's a couple of other containers running, and they don't seem to be having any problems reading and writing to/from the shared drives. (Well, one was having a problem but a quick chown 1000:1000 on that directory fixed it.)

**Reply by captainkev76** [reply]
Still no joy with this. I've just tried spinning up a fresh Ubuntu 20.04 install on an old PC I had, hoping to see if the problem is unique to Virtualbox. But the same problem exists on a vanilla Ubuntu server installed on a Celeron processor. I've tried mounting the drives as network shares. As before, all other containers work fine, but mariadb still giving permission errors and restarting every 60 seconds.

I think I've ran out of ideas now. Maybe time to admit defeat.

**Reply by lastzero** [reply]
Think I got it: https://mariadb.com/kb/en/server-system-variables/#lower_case_table_names

You probably need to append `--lower-case-table-names=1` to the mysqld command in your `docker-compose.yml`: 

```
command: mysqld --lower-case-table-names=1 --transaction-isolation=READ-COMMITTED --character-set-server=utf8mb4 --collation-server=utf8mb4_unicode_ci --max-connections=512 --innodb-rollback-on-timeout=OFF --innodb-lock-wait-timeout=120
```

Should have looked more closely at the filename (lower-test!) in your error. Most likely an issue with the case sensitivity of the Windows file system. Don't think Microsoft even offers case-sensitive file systems!?

**Reply by captainkev76** [reply]
Thanks for that - was worth a shot, but doesn't seem to have made a difference. I suspect that the lower-test file is just the first file it tries to create?

The other messages all definitely point towards a permissions issue (see below). I can't help thinking it's something to do with it "switching to dedicated user mysql", as that username doesn't exist outside of the container, so can't be added to the correct GID to have access. And the UID number it seems to use inside the container is already used by something else (systemd-coredump). Having said all that - I'm clearly no expert in this field.

```
mariadb_1     | 2021-10-26 18:45:07+00:00 [Note] [Entrypoint]: Switching to dedicated user 'mysql'
mariadb_1     | 2021-10-26 18:45:07+00:00 [Note] [Entrypoint]: Entrypoint script for MariaDB Server 1:10.6.4+maria~focal started.
mariadb_1     | 2021-10-26 18:45:07+00:00 [Note] [Entrypoint]: Initializing database files
mariadb_1     | 2021-10-26 18:45:07 0 [Warning] Can't create test file /var/lib/mysql/7dc30f8c2fd0.lower-test
mariadb_1     | 2021-10-26 18:45:08 0 [ERROR] mariadbd: Can't create/write to file '/var/lib/mysql/aria_log_control' (Errcode: 13 "Permission denied")
mariadb_1     | 2021-10-26 18:45:08 0 [ERROR] mariadbd: Got error 'Can't create file' when trying to use aria control file '/var/lib/mysql/aria_log_control'
mariadb_1     | 2021-10-26 18:45:08 0 [ERROR] Plugin 'Aria' init function returned error.
mariadb_1     | 2021-10-26 18:45:08 0 [ERROR] Plugin 'Aria' registration as a STORAGE ENGINE failed.
mariadb_1     | 2021-10-26 18:45:08 0 [ERROR] InnoDB: Operating system error number 13 in a file operation.
mariadb_1     | 2021-10-26 18:45:08 0 [ERROR] InnoDB: The error means mariadbd does not have the access rights to the directory.
mariadb_1     | 2021-10-26 18:45:08 0 [ERROR] InnoDB: Operating system error number 13 in a file operation.
mariadb_1     | 2021-10-26 18:45:08 0 [ERROR] InnoDB: The error means mariadbd does not have the access rights to the directory.
mariadb_1     | 2021-10-26 18:45:08 0 [ERROR] InnoDB: Cannot open datafile './ibdata1'
mariadb_1     | 2021-10-26 18:45:08 0 [ERROR] InnoDB: Could not open or create the system tablespace.
```

**Reply by lastzero** [reply]
Added this to our Windows example: https://dl.photoprism.org/docker/windows/docker-compose.yml

Let us know if it works! I was aware MySQL behaves differently with case-insensitive filesystems, but not that it can cause errors on startup...

**Reply by lastzero** [reply]
Damn, was hoping that's the issue. However, I found lots of search results when googling for your error message(s). Just didn't have time to read everything... if you have a bit of time, you might find the answer: 

https://www.google.com/search?q=Can%27t+create%2Fwrite+to+file+%27%2Fvar%2Flib%2Fmysql%2Faria_log_control%27+mariadb+windows+vm

**Reply by lastzero** [reply]
Make sure to also test MariaDB 10.5:
```
image: mariadb:10.5
```

Recently changed this to 10.6 assuming it doesn't cause any issues.

**Reply by captainkev76** [reply]
So I think I've made some progress today: I added user: 1002:998 to the docker-compose file under the mariadb section, as user 1002 is the uid of a username that has read/write access in windows, and 998 is the group which has group access from the Linux side. With this, the permissions errors seem to disappear, but are replaced with the following (with Mariadb10.5 and 10.6):

```
mariadb_1     | 2021-10-27 11:11:49 0x7fe120ffd700  InnoDB: Assertion failure in file /home/buildbot/buildbot/build/mariadb-10.6.4/storage/innobase/os/os0file.cc line 3594
mariadb_1     | InnoDB: Failing assertion: cb->m_err == DB_SUCCESS
mariadb_1     | InnoDB: We intentionally generate a memory trap.
mariadb_1     | InnoDB: Submit a detailed bug report to https://jira.mariadb.org/
mariadb_1     | InnoDB: If you get repeated assertion failures or crashes, even
mariadb_1     | InnoDB: immediately after the mariadbd startup, there may be
mariadb_1     | InnoDB: corruption in the InnoDB tablespace. Please refer to
mariadb_1     | InnoDB: https://mariadb.com/kb/en/library/innodb-recovery-modes/
mariadb_1     | InnoDB: about forcing recovery.
mariadb_1     | 211027 11:11:49 [ERROR] mysqld got signal 6 ;
mariadb_1     | This could be because you hit a bug. It is also possible that this binary
mariadb_1     | or one of the libraries it was linked against is corrupt, improperly built,
mariadb_1     | or misconfigured. This error can also be caused by malfunctioning hardware.
mariadb_1     |
mariadb_1     | To report this bug, see https://mariadb.com/kb/en/reporting-bugs
mariadb_1     |
mariadb_1     | We will try our best to scrape up some info that will hopefully help
mariadb_1     | diagnose the problem, but since we have already crashed,
mariadb_1     | something is definitely wrong and this may fail.
mariadb_1     |
mariadb_1     | Server version: 10.6.4-MariaDB-1:10.6.4+maria~focal
mariadb_1     | key_buffer_size=134217728
mariadb_1     | read_buffer_size=131072
mariadb_1     | max_used_connections=0
mariadb_1     | max_threads=514
mariadb_1     | thread_count=0
mariadb_1     | It is possible that mysqld could use up to
mariadb_1     | key_buffer_size + (read_buffer_size + sort_buffer_size)*max_threads = 1262890 K  bytes of memory
mariadb_1     | Hope that's ok; if not, decrease some variables in the equation.
mariadb_1     |
mariadb_1     | Thread pointer: 0x0
mariadb_1     | Attempting backtrace. You can use the following information to find out
mariadb_1     | where mysqld died. If you see no messages after this, something went
mariadb_1     | terribly wrong...
mariadb_1     | 2021-10-27 11:11:49 0x7fe1217fe700  InnoDB: Assertion failure in file /home/buildbot/buildbot/build/mariadb-10.6.4/storage/innobase/os/os0file.cc line 3594
mariadb_1     | InnoDB: Failing assertion: cb->m_err == DB_SUCCESS
mariadb_1     | InnoDB: We intentionally generate a memory trap.
mariadb_1     | InnoDB: Submit a detailed bug report to https://jira.mariadb.org/
mariadb_1     | InnoDB: If you get repeated assertion failures or crashes, even
mariadb_1     | InnoDB: immediately after the mariadbd startup, there may be
mariadb_1     | InnoDB: corruption in the InnoDB tablespace. Please refer to
mariadb_1     | InnoDB: https://mariadb.com/kb/en/library/innodb-recovery-modes/
mariadb_1     | InnoDB: about forcing recovery.
mariadb_1     | Aborted (core dumped)
mariadb_1     |
mariadb_1     | Installation of system tables failed!  Examine the logs in
mariadb_1     | /var/lib/mysql/ for more information.
mariadb_1     |
mariadb_1     | The problem could be conflicting information in an external
mariadb_1     | my.cnf files. You can ignore these by doing:
mariadb_1     |
mariadb_1     |     shell> /usr/bin/mysql_install_db --defaults-file=~/.my.cnf
mariadb_1     |
mariadb_1     | You can also try to start the mysqld daemon with:
mariadb_1     |
mariadb_1     |     shell> /usr/sbin/mariadbd --skip-grant-tables --general-log &
mariadb_1     |
mariadb_1     | and use the command line tool /usr/bin/mariadb
mariadb_1     | to connect to the mysql database and look at the grant tables:
mariadb_1     |
mariadb_1     |     shell> /usr/bin/mysql -u root mysql
mariadb_1     |     mysql> show tables;
mariadb_1     |
mariadb_1     | Try 'mysqld --help' if you have problems with paths.  Using
mariadb_1     | --general-log gives you a log in /var/lib/mysql/ that may be helpful.
mariadb_1     |
mariadb_1     | The latest information about mysql_install_db is available at
mariadb_1     | https://mariadb.com/kb/en/installing-system-tables-mysql_install_db
mariadb_1     | You can find the latest source at https://downloads.mariadb.org and
mariadb_1     | the maria-discuss email list at https://launchpad.net/~maria-discuss
mariadb_1     |
mariadb_1     | Please check all of the above before submitting a bug report
mariadb_1     | at https://mariadb.org/jira
mariadb_1     |
mariadb_1     | 2021-10-27 11:11:51+00:00 [Note] [Entrypoint]: Entrypoint script for MariaDB Server 1:10.6.4+maria~focal started.
mariadb_1     | 2021-10-27 11:11:51+00:00 [Note] [Entrypoint]: Initializing database files
mariadb_1     | 2021-10-27 11:11:52 0 [ERROR] InnoDB: Corrupted page [page id: space=0, page number=0] of datafile './ibdata1' could not be found in the doublewrite buffer.
mariadb_1     | 2021-10-27 11:11:52 0 [ERROR] InnoDB: Plugin initialization aborted with error Data structure corruption
mariadb_1     | 2021-10-27 11:11:52 0 [ERROR] Plugin 'InnoDB' init function returned error.
mariadb_1     | 2021-10-27 11:11:52 0 [ERROR] Plugin 'InnoDB' registration as a STORAGE ENGINE failed.
mariadb_1     | 2021-10-27 11:11:52 0 [ERROR] Unknown/unsupported storage engine: InnoDB
mariadb_1     | 2021-10-27 11:11:52 0 [ERROR] Aborting

```

**Comment by graciousgrey** [comment]
Did you went through the checklists here: https://docs.photoprism.org/getting-started/troubleshooting/?
